File: Python/230.Kth Smallest Element in a BST.py
#返回二叉搜索树中第K小的数
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        stack , p , i = [] , root , 1
        while p:
            stack.append(p)
            p = p.left
        while i!=k:
            p = stack.pop().right
            while p:
                stack.append(p)
                p = p.left
            i += 1
        return stack[-1].val

t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(5)
t6 = TreeNode(6)
t7 = TreeNode(7)
t8 = TreeNode(8)
t9 = TreeNode(9)
t6.left = t4
t6.right = t8
t4.left = t2
t4.right = t5
t2.left = t1
t2.right = t3
t8.left =  t7
t8.right = t9

s = Solution()
print(s.kthSmallest(t6,9))

File: C++/Tree/230.Kth Smallest Element in a BST.cpp
/*
返回二叉搜索树中第k小的数。
*/
/*
思路：二叉搜索树中序遍历是有序的。
使用中序非递归遍历找到第k个即可。
*/
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> mystack;
        TreeNode* p = root;
        int count = 0;
        while(p || !mystack.empty())
        {
        	while(p)
        	{
        		mystack.push(p);
        		p = p->left;
        	}
        	if(!mystack.empty())
        	{
        		p = mystack.top();
        		++count;
        		if(count == k)	return p->val;
        		mystack.pop();
        		p = p->right;
        	}
        }
        return -1;
    }
};

int main()
{
	Solution s;
	TreeNode n1(1),n2(2),n3(3),n4(4),n5(5),n6(6);
	n4.left = &n3, n4.right = &n5;
	n3.left = &n1, n1.right = &n2;
	n5.right = &n6;
	cout<<s.countNodes(&n4)<<endl;
}

